### PR TITLE
Type operator aliases

### LINKED ISSUE #416: Infix types
Is there a way to declare infix types as follows?

```purescript
data a :+: b = Inl a | Inr b

type f ~> g = forall a. f a -> g a
```

Thanks!

### PR DESCRIPTION
**Pull request by garyb**
Resolves #416

### CODE REVIEW

File: tests/TestDocs.hs [hunk 1]
```diff
@@ -285,12 +285,12 @@ testCases =
...
-        , TypeSynonymShouldRenderAs (n "ConstrainedArgument") "WithArgs" "forall a. (Foo a => a) -> a"
-        , TypeSynonymShouldRenderAs (n "ConstrainedArgument") "MultiWithoutArgs" "forall a. ((Partial, Partial) => a) -> a"
-        , TypeSynonymShouldRenderAs (n "ConstrainedArgument") "MultiWithArgs" "forall a b. ((Foo a, Foo b) => a) -> a"
-        ])
+    --, ("ConstrainedArgument",
+    --    [ TypeSynonymShouldRenderAs (n "ConstrainedArgument") "WithoutArgs" "forall a. (Partial => a) -> a"
+    --    , TypeSynonymShouldRenderAs (n "ConstrainedArgument") "WithArgs" "forall a. (Foo a => a) -> a"
+    --    , TypeSynonymShouldRenderAs (n "ConstrainedArgument") "MultiWithoutArgs" "forall a. ((Partial, Partial) => a) -> a"
+    --    , TypeSynonymShouldRenderAs (n "ConstrainedArgument") "MultiWithArgs" "forall a b. ((Foo a, Foo b) => a) -> a"
+    --    ])
```
Comment (garyb) [inline tests/TestDocs.hs]: I'm going to need a hand updating the pretty printer for types, as I have _no_ idea what's going on in there. Without it though, these tests spin, due to the `ParensInType` case not being covered.

**Comment by garyb** [comment]
Ok, this is ready for review and whatever. I still haven't managed to update the pretty printer though. I tried handling it by adding new cases to `typeLiterals` for `TypeOp`, `BinaryNoParensType` and `ParensInType `, but the doc tests still got stuck when encountering those that I commented out.

Note for @hdgarrood for when this goes in: I broke the externs format again, sorry.

**Comment by hdgarrood** [comment]
Thanks for the heads up. It's not really much of a problem for me now that I have that script for re-uploading everything, so please feel free to break the format whenever it's convenient. As you're probably already aware I'm a big proponent of getting the breaking changes in while we still can :)

**Comment by paf31** [comment]
Are we going to deprecate non-`type` imports like we did with `class`? I quite like the idea of explicit `data` imports too, now that I think about it.

Can we have unicode type operators?

**Comment by garyb** [comment]
I don't know... I had to introduce that keyword as our type operators have the same namespace as normal operators. In haskell all type operators have to be prefixed with `:`.

I wouldn't mind introducing a consistent keyword, but I'm not really pro-making-type-vs-data explicit. It seems like something you shouldn't necessarily need to know, and then also I guess `newtype` would fall under `data`, so it's not like we'll have an entirely consistent scheme for that. I'd prefer we just use `type` for anything, and have it behave the same was as the current `ProperName` imports; that is an export list can be provided or omitted.

And yep, unicode operators are fine! It's exactly the same as with operators at the value level.

**Comment by paf31** [comment]
That makes sense. I'm fine with using an extra keyword to disambiguate the namespaces.

**Comment by paf31** [comment]
Wait, this is merged, but it's missing the pretty printers?

**Comment by garyb** [comment]
Yes, I just realised that, sorry. I'm looking at it again now, although any tips would be appreciated.

**Pull request merged**